File: java-project-euler-100/src/main/java/net/talaatharb/projecteuler100/P020.java
package net.talaatharb.projecteuler100;

import java.math.BigInteger;
import java.util.HashMap;
import java.util.Map;

/**
 * P020
 * 
 * @author mharb
 *
 */
public class P020 extends Solution {

	/**
	 * Problem solution entry point
	 * 
	 * @param args
	 */
	public static void main(String[] args) {
		new P020().run();
	}

	private final Map<Integer, BigInteger> memoryCache;

	/**
	 * Public constructor
	 */
	public P020() {
		memoryCache = new HashMap<>();
		memoryCache.put(0, BigInteger.valueOf(1));
	}

	private int digitSum(final String string) {
		int result = 0;
		final char[] digits = string.toCharArray();
		for (char i : digits) {
			result += Character.getNumericValue(i);
		}

		return result;
	}

	private BigInteger factorial(final int n) {
		if (memoryCache.containsKey(n)) {
			return memoryCache.get(n);
		}
		final BigInteger data = BigInteger.valueOf(n).multiply(factorial(n - 1));
		memoryCache.put(n, data);
		return data;
	}

	@Override
	public Number solve() {
		problemNumber = 20;
		final int n = 100;

		for (int i = 1; i < n; i++) {
			factorial(i);
		}

		return digitSum(factorial(n).toString());
	}

}

File: typescript-project-euler-100/src/p020.ts
import { Solution } from "./Solution";

export class P020 extends Solution {

	digitSum(data: string): number {
		let result = 0;
		for (const digit of data) {
			result += Number(digit);
		}
		return result;
	}

	factorial(n: number): bigint {
		let result = BigInt(1);
		for (let i = 2; i <= n; i++) {
			result *= BigInt(i);
		}
		return result;
	}

    solve() {
        this.problemNumber = 20;
		const n = 100;
		const result = this.digitSum(this.factorial(n).toString());

		return result;
    }
}
